Release 4.0.0

wyattjoh opened the pull request:
# Release 4.0.0

## What does this PR do?
- Includes locale improvements from #1045 

### Client
- User Suspension and Banning Detailed Info - In the User Detail Drawer
- Ability to perform actions to the User Detail Drawer such as: Ban, Suspend, Remove Ban and Remove Suspension. 
- The Community Section now relies on the graph. All the REST endpoints were deprecated.
- New Features: People in the Community Section for actions over User Status and Roles.

Bugs fixing
- Moderation Actions overlaps drawer - https://www.pivotaltracker.com/story/show/154035508
- Moment.js settings under user's locale
- UI reducer to manage UI state


_TODO: Summary under construction_

wyattjoh committed:
added support for user assigned_by fields to get resolved

wyattjoh committed:
added role management to graph

wyattjoh committed:
removed double await

wyattjoh committed:
fixes based on feedback

okbel committed:
missing fragment

okbel committed:
Adding diff and fix to utils

okbel committed:
Working account history

okbel committed:
Build User History

okbel committed:
UserDetail Refactor

okbel committed:
UserDetail Refactor

okbel committed:
Adding addUserRole and removeUserRole

okbel committed:
Adding UserDetailCommentList

okbel committed:
Approve Username fix

wyattjoh committed:
roles -> role, null -> COMMENTER for user roles

wyattjoh committed:
fixed bug with featuring comments doesn't work

wyattjoh committed:
more role fixes

okbel committed:
Adding ban message and setRole

okbel committed:
Ban user dialog steps

wyattjoh committed:
Merge branch 'master' into next

wyattjoh committed:
Merge branch 'next' into user-status-refactor

okbel committed:
Translations and Styling

okbel committed:
Merge branch 'user-status-refactor' of github.com:coralproject/talk into user-status-refactor

okbel committed:
Wrong filename

okbel committed:
Removing ternary

okbel committed:
Updating after approving a user

okbel committed:
Linting

wyattjoh committed:
Merge branch 'master' into next

kgardnr committed:
Merge branch 'master' into next

okbel committed:
Merge branch 'master' of github.com:coralproject/talk into user-status-refactor

okbel committed:
New path

okbel committed:
Updated comment user status path

wyattjoh committed:
added better compression support

wyattjoh committed:
Merge branch 'next' into user-status-refactor

wyattjoh committed:
improved query performance with indexes, added engine for debugging

okbel committed:
Moving ActionsMenu

okbel committed:
Styling and functionality

okbel committed:
Functionality and styling

okbel committed:
UserInfoTooltip, styling and func

okbel committed:
Adding old user status

okbel committed:
Adding old user status

okbel committed:
wip

okbel committed:
Adding react-mdl

okbel committed:
Updated tests and Drawer component

wyattjoh committed:
Adjusted mongo indexes to support index scanning

alinhle commented:
Code looks good 👍

wyattjoh committed:
Merge branch 'master' into next

wyattjoh committed:
Merge branch 'next' into user-status-refactor

kgardnr committed:
Update CSS classname

wyattjoh committed:
added support for deep route caching, settings cache, template optim

wyattjoh committed:
removed reference

wyattjoh committed:
fixed small bug

wyattjoh committed:
Merge branch 'master' into next

wyattjoh committed:
Merge branch 'next' into user-status-refactor

okbel committed:
Merge branch 'actions-user-drawer' of github.com:coralproject/talk into user-status-refactor

okbel committed:
merge

wyattjoh committed:
fixed template errors

wyattjoh committed:
linting patch

okbel committed:
Updated with User-Status-Refactor API

okbel committed:
Adding unSuspendUser mutation

okbel committed:
Updated with User-Status-Refactor API

okbel committed:
Updated with User-Status-Refactor API - User Detail Updated

okbel committed:
unBan Update

okbel committed:
Adding utils, and replacing REST old endpoints

okbel committed:
Styling, refactor and instant update after banning and unbanning from community

okbel committed:
Avoiding refetch from Community

wyattjoh committed:
Fixed import error

wyattjoh committed:
Merge branch 'master' into next

wyattjoh committed:
Merge branch 'next' into user-status-refactor

kgardnr committed:
Merge branch 'master' into next

kgardnr committed:
Merge branch 'master' into next

okbel committed:
merge conficts

okbel committed:
Adding role, instead of roles. And disable banning and suspending for the user thats logged in

okbel committed:
Classes for AccountHistory

okbel committed:
Adding Classes to the PageObject

okbel committed:
Test

okbel committed:
Adding Classes

okbel committed:
Updating the store mutations

okbel committed:
Adding Classes

okbel committed:
Updating the store mutations

okbel committed:
Adding tests

okbel committed:
Updated tests

wyattjoh committed:
removed dev requirement for engine

wyattjoh committed:
linting

okbel committed:
rejected usernames should not be able to comment

wyattjoh committed:
added mount path to the endpoint to respect the mounting config

okbel committed:
Updating the store mutations

okbel committed:
bounding fns

okbel committed:
Adding tests

okbel committed:
Merge branch 'user-status-refactor' of github.com:coralproject/talk into user-status-refactor

* 'user-status-refactor' of github.com:coralproject/talk:
  added mount path to the endpoint to respect the mounting config
  linting
  removed dev requirement for engine

okbel committed:
updated tests classes

kgardnr committed:
Merge branch 'master' of git+ssh://github.com/coralproject/talk into user-status-refactor

wyattjoh committed:
Merge branch 'master' into next

wyattjoh committed:
Merge branch 'next' into user-status-refactor

wyattjoh committed:
fixed graphql error

wyattjoh committed:
hid not logged in error

wyattjoh committed:
fixed dist loading error

wyattjoh committed:
updates to cli

- removed -c,--config flag in favor of env only or .env files
- loads .env always if found
- scripts re-written to call cli-serve directly

wyattjoh committed:
fixed tests

wyattjoh committed:
fixed tests

wyattjoh committed:
Merge pull request #1154 from coralproject/user-status-refactor

User status refactor

wyattjoh committed:
Merge branch 'master' into next

kgardnr committed:
Fix merge conflict

alinhle commented on bin/cli-users:
```diff
@@ -37,11 +37,11 @@ function getUserCreateAnswers(options) {
 
     if (options.role && USER_ROLES.indexOf(options.role) > -1) {
-      user.roles = [options.role];
+      user.roles = options.role;
```
Shouldn't this be `user.role`?

alinhle commented on bin/commander.js:
```diff
@@ -1,51 +0,0 @@
-  util.pid(parseArgs.pid);
-}
-
-module.exports = program
```
Why was this file removed? I used to do e.g. `./bin/cli -c .env users` but now the parameter config is no longer supported?

alinhle replied:
Also no more version and `pid` parameter

wyattjoh replied:
Now the `.env` file is read if you're in development. It complicated the cli, and this way it _just works_ easier.

okbel committed:
Merge branch 'master' into next

alinhle commented on bin/verifications/database/action_counts.js:
```diff
@@ -0,0 +1,185 @@
+
+      // And we generate the group id.
+      const ACTION_TYPE = sc(actionSummary.action_type.toLowerCase());
+      const GROUP_ID = sc(actionSummary.group_id.toLowerCase());
```
Aren't `ACTION_TYPE` and `GROUP_ID` already in snake case?

alinhle commented on bin/verifications/database/action_counts.js:
```diff
@@ -0,0 +1,185 @@
+      if (
+        !document.action_counts ||
+        !(ACTION_COUNT_FIELD in document.action_counts) ||
+        document.action_counts[ACTION_COUNT_FIELD] !== actionSummary.count
```
This can be simplified using lodash:
`get(document, ['action_counts', ACTION_COUNT_FIELD]) !== actionSummary.count`

alinhle commented on bin/verifications/database/action_counts.js:
```diff
@@ -0,0 +1,185 @@
+
+    // Group all the action summaries together from all the different group
+    // ids.
+    let groupedActionSummaries = actionSummaries.reduce((acc, actionSummary) => {
```
Should be `const` I think

alinhle replied:
Is linter activated for these files?

wyattjoh replied:
The linter is activated, but we haven't enabled the [prefer-const](https://eslint.org/docs/rules/prefer-const) rule in our [.eslintrc.json](https://github.com/coralproject/eslint-config-talk/blob/master/server.js).

alinhle commented on bin/verifications/database/action_counts.js:
```diff
@@ -0,0 +1,185 @@
+    // Group all the action summaries together from all the different group
+    // ids.
+    let groupedActionSummaries = actionSummaries.reduce((acc, actionSummary) => {
+      const ACTION_TYPE = sc(actionSummary.action_type.toLowerCase());
```
action_type should be already in snake case?

alinhle commented on bin/verifications/database/action_counts.js:
```diff
@@ -0,0 +1,185 @@
+      if (
+        !document.action_counts ||
+        !(ACTION_COUNT_FIELD in document.action_counts) ||
+        document.action_counts[ACTION_COUNT_FIELD] !== count
```
Lodash.get can simplify this

alinhle commented on client/coral-admin/src/components/AccountHistory.js:
```diff
@@ -0,0 +1,63 @@
+          </div>
+          {
+            userHistory.map((h, i) => (
+              <div className={cn(styles.row, 'talk-admin-account-history-row')} key={i}>
```
Needs a stable `key` instead of using `index`

okbel committed:
fix index for moderation actions, and view options

wyattjoh committed:
added user query by value to graph

wyattjoh committed:
Merge branch 'master' into next

wyattjoh committed:
Merge branch 'next' into user-search

alinhle commented:
Load More issues: 
![image](https://user-images.githubusercontent.com/14221600/34564191-0231de62-f156-11e7-8708-be233348c32a.png)

alinhle commented on client/coral-admin/src/components/UserDetail.js:
```diff
@@ -82,13 +63,19 @@ export default class UserDetail extends React.Component {
-  showAll = () => {
-    this.props.changeStatus('all');
+  changeTab = (tab) => {
+    this.props.changeStatus(tab);
```
We should change the prop name from `changeStatus` to `changeTab`, also related props in the container and redux actions to make it more clear.

alinhle commented on client/coral-admin/src/components/UserDetail.js:
```diff
@@ -100,36 +87,90 @@ export default class UserDetail extends React.Component {
-      toggleSelectAll
+      toggleSelectAll,
+      unBanUser,
+      unSuspendUser,
```
I'd call them `unbanUser` and `unsuspendUser` as `unban` and `unsuspend` is one word.

alinhle commented on client/coral-admin/src/components/UserDetail.js:
```diff
@@ -100,36 +87,90 @@ export default class UserDetail extends React.Component {
     let rejectedPercent = (rejectedComments / totalComments) * 100;
-    if (rejectedPercent === Infinity || isNaN(rejectedPercent)) {
 
-      // if totalComments is 0, you're dividing by zero, which is naughty
```
Keep the comment?

okbel replied:
Done

alinhle commented on client/coral-admin/src/components/UserDetail.js:
```diff
@@ -100,36 +87,90 @@ export default class UserDetail extends React.Component {
+    const suspended = isSuspended(user);
+    
     return (
       <ClickOutside onClickOutside={hideUserDetail}>
```
Need some logic here to only hide when no dialog is shown

okbel replied:
Which dialog? The drawer?

alinhle replied:
I'd add a new reducer called `ui` with one property `modal: Boolean` and pass that to this component e.g. `onClickOutside={!modal && hideUserDetail}`

alinhle replied:
`modal` should be true whenever a dialog is shown (BanDialog, SuspendDialog). The reducer can just act upon the actions from BanDialog and SuspendDialog

alinhle replied:
and false when they dissappear

okbel replied:
Done :D

alinhle commented on client/coral-admin/src/components/UserDetail.js:
```diff
@@ -197,41 +260,55 @@ export default class UserDetail extends React.Component {
+                  }} />
+                <label htmlFor='toogleAll'>Select all</label>
+              </div>
+            )}
```
We should move this part inside `UserDetailCommentList`

alinhle replied:
All that logic with selecting and bulk actions shouldn't be here anymore

alinhle commented:
Weird locals for the date 
![image](https://user-images.githubusercontent.com/14221600/34565365-e001fb42-f15a-11e7-995c-02f998c3d684.png)

okbel committed:
Merge pull request #1244 from coralproject/user-search

User Search

okbel committed:
Merge branch 'master' of github.com:coralproject/talk into next

* 'master' of github.com:coralproject/talk:
  fixes default lang select bug

okbel committed:
Merge branch 'next' of github.com:coralproject/talk into next

* 'next' of github.com:coralproject/talk:
  added user query by value to graph

alinhle commented on client/coral-admin/src/graphql/index.js:
```diff
@@ -1,35 +1,157 @@
   mutations: {
-    SetUserStatus: ({variables: {status, userId}}) => ({
+    SetUserRole: ({variables: {id: userId, role}}) => ({
       updateQueries: {
```
Better to use the DataProxy here

alinhle replied:
Good job 👍

alinhle commented:
Missing sorting and filtering functionality
![image](https://user-images.githubusercontent.com/14221600/34568071-e58afbfe-f164-11e7-89d1-1ac71b73bf4a.png)

alinhle commented on client/coral-admin/src/routes/Community/components/People.js:
```diff
@@ -1,75 +1,190 @@
+                    </thead>
+                    <tbody>
+                      {users.nodes.map((user, i)=> (
+                        <tr key={i} className="talk-admin-community-people-row">
```
Use key: `user.id`

okbel replied:
Done

okbel committed:
Search functionality on the Community People Section

okbel committed:
Using user.id

okbel committed:
Naming conventions

okbel committed:
Naming Conventions

okbel committed:
SetUserRole using DataProxy

okbel committed:
Account History stable key

alinhle commented on client/coral-admin/src/routes/Community/components/RejectUsernameDialog.js:
```diff
@@ -79,28 +79,28 @@ class RejectUsernameDialog extends Component  {
                 </div>
-        }
-        <div className={cn(styles.modalButtons, 'talk-reject-username-dialog-buttons')}>
+        } */}
```
We don't need this anymore or why was that commented out?

okbel replied:
This functionality it's not entirely done on the BE and it's not going to be released in this version.

alinhle replied:
Add some comments there to clarify

okbel replied:
👍

alinhle commented on client/coral-admin/src/routes/Community/components/styles.css:
```diff
@@ -240,3 +240,97 @@
 
+/* ========================================================================== */
+/* TABLE */
+/* ========================================================================== */
```
`styles.css` is deprecated, we shouldn't add more styles to it.

okbel replied:
👍 Adding a comment then.

okbel committed:
Moving bulk actions to User tab

alinhle commented:
Flagged User pagination is not working
![image](https://user-images.githubusercontent.com/14221600/34572133-de722f70-f170-11e7-98ea-80bfe8a0e26a.png)

alinhle commented on client/coral-admin/src/routes/Community/containers/People.js:
```diff
@@ -1,104 +1,164 @@
+        const updated = update(previous, {
+          flaggedUsers: {
+            nodes: {
+              $apply: (nodes) => appendNewNodes(nodes, users.nodes),
```
Why not just: `{$push: users.nodes}` ?

alinhle commented on client/coral-admin/src/routes/Community/containers/People.js:
```diff
@@ -1,104 +1,164 @@
       e.preventDefault();
-      this.fetchUsers();
+
+      // this.fetchUsers();
```
Just remove it completely?

alinhle commented on client/coral-admin/src/routes/Community/containers/People.js:
```diff
@@ -1,104 +1,164 @@
-    this.fetchUsers();
+    // this.timer = setTimeout(() => {
+    //   // this.fetchUsers({value});
+    // }, 350);
```
Why is this commented out? Additional info would be helpful

alinhle commented on client/coral-admin/src/routes/Community/containers/People.js:
```diff
@@ -1,104 +1,164 @@
 
   onSearchChange = (e) => {
     const value = e.target.value;
+    console.log(value);
```
^ console.log

okbel commented:
Please pull the latest, you are reviewing old code @cvle

alinhle commented on client/coral-admin/src/routes/Community/containers/People.js:
```diff
@@ -1,104 +1,230 @@
+        const updated = update(previous, {
+          users: {
+            nodes: {
+              $apply: (nodes) => appendNewNodes(nodes, users.nodes),
```
Why not just `${push: users.nodes}`?

alinhle replied:
`appendNewNodes` only appends new nodes as the name suggests and to answer my own question :-P

okbel replied:
_Solved via Slack_

kgardnr committed:
Merge branch 'master' into next

alinhle commented:
Table header turns cursor into `pointer` but it does not trigger any action.
If we don't have sorting support, remove the pointer cursor styling.

![image](https://user-images.githubusercontent.com/14221600/34573148-a289db36-f173-11e7-9541-eae0c27ea5f3.png)

alinhle commented on client/coral-admin/src/routes/Configure/components/EmbedLink.js:
```diff
@@ -5,6 +5,7 @@ import styles from './EmbedLink.css';
 import {Button} from 'coral-ui';
 import {BASE_URL} from 'coral-framework/constants/url';
 import ConfigureCard from 'coral-framework/components/ConfigureCard';
+import {stripIndent} from 'common-tags';
```
I love this

okbel committed:
Removing pointer hover, adding proptypes

alinhle commented on client/coral-admin/src/routes/Moderation/containers/Comment.js:
```diff
@@ -37,7 +37,19 @@ export default withFragments({
         id
         username
-        status
+        state {
```
I think these are unused. I'll verify this later on.

alinhle replied:
Seems so, let's remove it.

alinhle commented on client/coral-framework/graphql/mutations.js:
```diff
@@ -177,39 +177,135 @@ export const withSuspendUser = withMutation(
+    }),
+  });
+
+export const withUnBanUser = withMutation(
```
Should be `UnbanUser`. We can do a simple Search and Replace throughout the project. As well as for `UnsuspendUser`

okbel replied:
Done (:

okbel replied:
_I don't know why GH is not removing this_

alinhle commented on client/coral-configure/components/CloseCommentsInfo.js:
```diff
@@ -0,0 +1,29 @@
+  onClick: PropTypes.func,
+};
+
+export default CloseCommentsInfo;
```
What's up with thise file. It's not used?

okbel replied:
Maybe I pushed it back 🤔 Checking...

okbel replied:
Done.

okbel committed:
Naming Conventions

okbel committed:
Merge branch 'next' of github.com:coralproject/talk into next

* 'next' of github.com:coralproject/talk:
  Up patch version to match tag.

alinhle commented on client/coral-embed-stream/src/components/ChangeUsername.js:
```diff
@@ -0,0 +1,76 @@
+          placeholder={t('framework.edit_name.label')}
+          id='username'
+          onChange={(e) => this.setState({username: e.target.value})}
+          rows={3}/><br/>
```
Why does it say `rows=3`?

okbel replied:
No idea. Gone.

alinhle commented on client/coral-embed-stream/src/components/ChangeUsername.js:
```diff
@@ -0,0 +1,76 @@
+          className={cn(styles.editNameInput, 'talk-change-username-username-input')}
+          value={username}
+          placeholder={t('framework.edit_name.label')}
+          id='username'
```
what's the `id` for?

okbel replied:
Removing.

alinhle commented on client/coral-embed-stream/src/components/ChangeUsername.js:
```diff
@@ -0,0 +1,76 @@
+        <label
+          htmlFor='username'
+          className="screen-reader-text"
+          aria-hidden={true}>
```
Why is it hidden from screen readers using  `aria-hidden=true` when it is meant for screen readers? 🤔

okbel replied:
No idea.

okbel replied:
Removing it

okbel replied:
Removing.

alinhle commented on client/coral-embed-stream/src/reducers/auth.js:
```diff
@@ -228,7 +228,7 @@ export default function auth (state = initialState, action) {
     };
   case 'APOLLO_SUBSCRIPTION_RESULT':
-    if (action.operationName === 'UserBanned' && state.getIn(['user', 'id']) === action.variables.user_id) {
+    if (action.operationName === 'UserBanned' && state.user.id === action.variables.user_id) {
```
Add `@TODO: These don't work anymore because apollo store has been decoupled`.

okbel replied:
👍

okbel committed:
Adding warning

okbel committed:
Removing unused code

okbel committed:
Removing unused code

alinhle commented on client/coral-framework/services/i18n.js:
```diff
@@ -3,6 +3,11 @@ import has from 'lodash/has';
+import 'moment/locale/da';
+import 'moment/locale/es';
+import 'moment/locale/fr';
+import 'moment/locale/pt-br';
```
Missing english locales :-P (I guess)

okbel replied:
Nope haha! We needed to specify moment's locale. Done! 👍

alinhle commented on client/coral-ui/components/Drawer.js:
```diff
@@ -1,11 +1,13 @@
-    <div className={styles.drawer}>
-      <div className={styles.closeButton} onClick={onClose}>×</div>
+    <div className={cn(styles.drawer, className)}>
+      {/* TODO: Swap out with button */}
```
Let's do it :-)

okbel replied:
I didn't add that line. It's meant for the Button Component? 🤔

alinhle replied:
No I think just the normal html button. I thought you added the line^^

okbel replied:
Done

alinhle commented on config.js:
```diff
@@ -7,6 +7,13 @@
+
+  // Apply all the configuration provided in the .env file if it isn't already
+  // in the environment.
+  require('dotenv').config();
```
Doesn't seem to work for me. Where does it look for a `.env` file?

alinhle commented on errors.js:
```diff
@@ -165,12 +165,6 @@ const ErrPermissionUpdateUsername = new APIError('You do not have permission to
 });
 
-// ErrSameUsernameProvided is returned when attempting to update a username to the same username.
-const ErrSameUsernameProvided = new APIError('Same username provided.', {
```
We don't enforce this on the server side? @wyattjoh

wyattjoh replied:
We do!

https://github.com/coralproject/talk/pull/1153/files#diff-79a167f71b013e1de67ca4450084bc9cR283

wyattjoh committed:
pr review updates

wyattjoh committed:
changed beheviour of dotenv parsing

wyattjoh committed:
fixed let -> const

kgardnr committed:
Merge branch 'master' into next

kgardnr commented:
@okbel and @cvle, I'm seeing a few errors on Admin:

- When I go to Admin>Moderate, I get `Error` and no moderation functionality shows

- When I go to Admin>Community I get `GraphQL error: Cannot return null for non-nullable field BannedStatus.status. GraphQL error: Cannot return null for non-nullable field BannedStatus.status.`

okbel commented:
@kgardnr It says that status is returning `null` and it shouldn't. Did you run the migrations?

okbel committed:
Removing unused code

okbel committed:
Moment settings based on users locale

okbel committed:
Load more: Rejected Usernames -> Wrong name

okbel committed:
Adding clarifications

alinhle commented on graph/loaders/users.js:
```diff
@@ -31,32 +73,59 @@ const genUserByIDs = async (context, ids) => {
+
+    if (action_type) {
+      query.merge({
+        [`action_counts.${sc(action_type.toLowerCase())}`]: {
```
also here, should be already in snake case.

alinhle commented on graph/loaders/users.js:
```diff
@@ -115,25 +185,25 @@ const getUsersByQuery = async ({user, loaders: {Actions}}, {ids, limit, cursor,
-    });
+    if (action_type) {
+      query.merge({
+        [`action_counts.${sc(action_type.toLowerCase())}`]: {
```
Already snake case

alinhle commented on plugins/talk-plugin-featured-comments/index.js:
```diff
@@ -57,27 +57,23 @@ module.exports = {
     RootMutation: {
       addTag: {
-        async post(obj, {tag: {name, id, item_type}}, {user, mutators: {Comment}, pubsub}, info, result) {
+        async post(obj, {tag: {name, id, item_type}}, {user, mutators: {Comment}, pubsub}, _info) {
```
Just remove `_info`?

alinhle commented on plugins/talk-plugin-featured-comments/index.js:
```diff
@@ -57,27 +57,23 @@ module.exports = {
       },
       removeTag: {
-        async post(obj, {tag: {name, id, item_type}}, {user, loaders: {Comments}, pubsub}, info, result) {
+        async post(obj, {tag: {name, id, item_type}}, {user, loaders: {Comments}, pubsub}, _info) {
```
Just remove `_info`?

alinhle commented on services/actions.js:
```diff
@@ -248,3 +251,56 @@ module.exports = class ActionsService {
 };
+
+const incrActionCounts = async (action, value) => {
+  const ACTION_TYPE = sc(action.action_type.toLowerCase());
```
Already in snake case?

alinhle commented on services/actions.js:
```diff
@@ -248,3 +251,56 @@ module.exports = class ActionsService {
+  };
+
+  if (action.group_id && action.group_id.length > 0) {
+    const GROUP_ID = sc(action.group_id.toLowerCase());
```
Already in snake case?

alinhle commented on services/users.js:
```diff
@@ -82,8 +83,246 @@ module.exports = class UsersService {
+      {
+        id,
+        status: {
+          $ne: status,
```
Is this query here correct? looks odd

wyattjoh replied:
This prevents a user (who is currently banned) from being banned again. As each time we push the item into history. Without, it would look like:

```js
[
  {status: BANNED},
  {status: BANNED},
  {status: ACTIVE}
]
```

Now we ensure that the user doesn't already have that status first.

alinhle replied:
But status is no longer a string, but an object of statuses

alinhle replied:
Should be sth like `{status: {banned: {status: status}}` no?

alinhle commented on services/users.js:
```diff
@@ -82,8 +83,246 @@ module.exports = class UsersService {
+      {
+        id,
+        status: {
+          $ne: status,
```
Is this query here correct? looks odd

wyattjoh replied:
See above.

alinhle replied:
See above :-)

alinhle commented on services/users.js:
```diff
@@ -82,8 +83,246 @@ module.exports = class UsersService {
+        }
+
+        if (!resetAllowed && user.username === username) {
+          throw errors.ErrUsernameTaken;
```
Did you kick out `ErrSameUsernameProvided` and use `ErrUsernameTaken` on purpose?

alinhle replied:
@wyattjoh

wyattjoh replied:
Yes. We used it inconsistently before, so this reduces the amount of variation that you get from it.

alinhle replied:
It just does not precisely describe the error situation here as it only occurs when I'm changing to my own username.

alinhle commented on test/e2e/page_objects/admin.js:
```diff
@@ -55,12 +55,16 @@ module.exports = {
+    usernameDialogStep1: '.talk-admin-reject-username-dialog-step-1',
+    usernameDialogButtons: '.talk-admin-reject-username-dialog-buttons',
+    usernameDialogSuspend: '.talk-admin-reject-username-dialog-button-k',
+    usernameDialogSuspensionMessage: '.talk-admin-reject-username-dialog-suspension-message',
```
Move these to a section

okbel replied:
😫

alinhle replied:
@okbel are you working on these?

alinhle commented on test/e2e/page_objects/admin.js:
```diff
@@ -55,12 +55,16 @@ module.exports = {
     signOutButton: '.talk-admin-header-sign-out',
     suspendUserDialog: '.talk-admin-suspend-user-dialog',
+    suspendUserDialogStep0: '.talk-admin-suspend-user-dialog-step-0',
+    suspendUserDialogStep1: '.talk-admin-suspend-user-dialog-step-1',
```
Move these to a section

okbel replied:
😩

alinhle commented on test/e2e/specs/04_userStatus.js:
```diff
@@ -1,7 +1,6 @@
 module.exports = {
 
   before: (client) => {
-    client.setWindowPosition(0, 0);
```
This is important to keep.

alinhle replied:
On browserstack the browser does not always start at 0, 0

okbel replied:
I think this was a merge confict :p I will add it back

kgardnr replied:
Oops that was me, sorry!

alinhle replied:
@okbel

alinhle commented on graph/typeDefs.graphql:
```diff
@@ -1278,17 +1390,38 @@ type RootMutation {
+  # Sets the ban status on a given user. Requires the `MODERATOR` role.
   # Mutation is restricted.
-  rejectUsername(input: RejectUsernameInput!): RejectUsernameResponse
+  unBanUser(input: UnBanUserInput!): UnBanUserResponse
```
@wyatt can we change them to `unbanUser` and `unsuspendUser`. My rules are to captilize whenever there is a new word, but `unBan` and `unSuspend` does not follow this rule. It's more complicated to say _capitalize whenever there is a new word UNLESS the word marks the opposite effect (undo) of another action in which case un is prefixed to the first-letter-capitalized name of the other action_ than just _capitalize whenever there is a new word_

wyattjoh replied:
It's likely my bad english ha, I'll apply a refactor to resolve that.

alinhle commented on config.js:
```diff
@@ -7,6 +7,10 @@
 
+// Apply all the configuration provided in the .env file if it isn't already
+// in the environment.
+require('dotenv').config();
```
Posting this again, cause it dissappeared into the outdated zone:

Doesn't seem to work for me. Where does it look for a `.env` file?

wyattjoh replied:
In your root talk directory, it will load configuration into the process if it isn't already in the environment.

alinhle replied:
Oh it didn't work for me before, but now it does.

okbel committed:
Adding line back

okbel committed:
Using button in the Drawer

okbel committed:
styles.css is de-pre-ca-ted

okbel committed:
Adding a UI reducer to manage UI state

alinhle committed:
Fix styling for new x button

alinhle committed:
Use null instead of boolean

alinhle committed:
Remove unused state

alinhle committed:
Mark onClickOutside not required

alinhle committed:
Add proptype

wyattjoh committed:
Merge branch 'master' into next

wyattjoh committed:
resolving pr concerns

wyattjoh committed:
fixed wrong query logic

wyattjoh committed:
reintroduced the errors.ErrSameUsernameProvided

wyattjoh committed:
applied rename to graph mutation edges

wyattjoh committed:
don't apply configuration from dotfiles during testing

wyattjoh committed:
user cli patches

- removed unused user actions
- added system side graph context
- added user cli search

wyattjoh committed:
hardened configuration

- moved env rewrite + dotenv code to single location
- fixed bug with system flag on user

alinhle committed:
Rename UnBan and UnSuspend

alinhle committed:
Remove defunct refetch detection workaround

alinhle committed:
Fix change username not displaying errors

alinhle committed:
Move ChangeUsername to the appropiate tab

alinhle committed:
Fix live status updates

wyattjoh committed:
refactored mailer service with lodash!

alinhle committed:
Use correct mutation and show errors

wyattjoh committed:
Merge branch 'next' into fix-live-status-changes

wyattjoh committed:
fix to optional required param

wyattjoh committed:
Merge branch 'master' into next

wyattjoh committed:
removed snakecase

wyattjoh committed:
Merge branch 'next' into fix-live-status-changes

wyattjoh committed:
Merge pull request #1255 from coralproject/fix-live-status-changes

Fix live status updates

alinhle committed:
Merge branch 'next' into fix-set-username

alinhle committed:
Fully implement SetUsername

okbel committed:
updated tests

okbel committed:
updated tests

okbel committed:
Merge branch 'next' of github.com:coralproject/talk into next

* 'next' of github.com:coralproject/talk: (26 commits)
  removed snakecase
  fix to optional required param
  refactored mailer service with lodash!
  Docs typo
  Fix live status updates
  Move ChangeUsername to the appropiate tab
  Fix change username not displaying errors
  Remove defunct refetch detection workaround
  Rename UnBan and UnSuspend
  hardened configuration
  user cli patches
  don't apply configuration from dotfiles during testing
  applied rename to graph mutation edges
  reintroduced the errors.ErrSameUsernameProvided
  fixed wrong query logic
  resolving pr concerns
  Add proptype
  Mark onClickOutside not required
  Remove unused state
  Use null instead of boolean
  ...

okbel committed:
typo

alinhle committed:
Merge branch 'next' into fix-set-username

kgardnr committed:
Merge branch 'master' into next

kgardnr committed:
Merge branch 'next' into fix-set-username

kgardnr committed:
Merge pull request #1259 from coralproject/fix-set-username

Make Facebook login work again in next

kgardnr committed:
Merge branch 'master' into next

alinhle committed:
Add some margin to loadmore

okbel committed:
Adding missing proptypes

okbel committed:
Removing replyButton in Stream is closed

okbel committed:
Adding e2e for the stream - open and closed asset features

okbel committed:
remove uncommitted

okbel committed:
Merge branch 'master' into reply

alinhle committed:
Use lodash with es modules

okbel committed:
Updated tests

okbel committed:
Merge branch 'reply' of github.com:coralproject/talk into reply

* 'reply' of github.com:coralproject/talk:
  Small grammar fixes
  refactor to cleanup
  extracted asset url
  extracted more funcs
  extracted events from opts
  Adding TALK_PERSPECTIVE_DO_NOT_STORE to the docs
  Adding doNotStore
  Translation update to make it more appropriate.
  Translation for hide off topic comments

okbel committed:
Updated tests

alinhle committed:
Merge branch 'next' into reply

okbel committed:
Merge branch 'reply' of github.com:coralproject/talk into reply

* 'reply' of github.com:coralproject/talk: (193 commits)
  Add some margin to loadmore
  typo
  updated tests
  updated tests
  Fully implement SetUsername
  removed snakecase
  fix to optional required param
  Use correct mutation and show errors
  refactored mailer service with lodash!
  Fix live status updates
  Move ChangeUsername to the appropiate tab
  Fix change username not displaying errors
  Remove defunct refetch detection workaround
  Rename UnBan and UnSuspend
  hardened configuration
  user cli patches
  don't apply configuration from dotfiles during testing
  applied rename to graph mutation edges
  reintroduced the errors.ErrSameUsernameProvided
  fixed wrong query logic
  ...

alinhle committed:
Fix typo

wyattjoh committed:
adjusted build script to build the next branch and release it

wyattjoh committed:
removed debug...

wyattjoh committed:
Merge pull request #1266 from coralproject/circle-next

Circle + Docker next

alinhle committed:
Merge branch 'next' into reply

alinhle committed:
Merge branch 'next' into lodash-size-opt

wyattjoh committed:
Merge pull request #1265 from coralproject/lodash-size-opt

Use lodash with es modules

okbel committed:
Merge branch 'next' into reply

wyattjoh committed:
Merge pull request #1264 from coralproject/reply

Hide reply button when comment stream is closed

wyattjoh committed:
fixed perms issue

okbel committed:
Icon User Drawer

okbel committed:
ignore user box over flow, fixed

okbel committed:
Updating copy

okbel committed:
Feature Button

wyattjoh committed:
Merge branch 'next' into perm-fix

wyattjoh committed:
changed copy for error

alinhle committed:
Fix suspect label

alinhle committed:
Merge branch 'next' into perm-fix

alinhle committed:
Merge branch 'next' into suspect-label

alinhle committed:
Fix no status history error

alinhle committed:
Remove unwanted scroll in configure

alinhle committed:
Only show scrollbars when necessary

alinhle committed:
Restore shortcut functionality

kgardnr committed:
Merge pull request #1270 from coralproject/suspect-label

Fix suspect label

kgardnr committed:
Merge branch 'next' into fix-no-status-history

alinhle committed:
Change Flags to Reports and make a bit bigger

okbel committed:
fixes

okbel committed:
Merge branch 'next' into account-history

alinhle committed:
Merge branch 'next' into flags-enh

wyattjoh committed:
added email fixes

kgardnr committed:
Merge pull request #1271 from coralproject/fix-no-status-history

Fix no status history error

kgardnr committed:
Merge branch 'next' into fix-scrollbars

kgardnr committed:
Merge pull request #1272 from coralproject/fix-scrollbars

Fix scrollbars issues

kgardnr committed:
Merge branch 'next' into selected-bug

kgardnr committed:
Merge pull request #1273 from coralproject/selected-bug

Restore shortcut functionality

kgardnr committed:
Merge branch 'next' into flags-enh

kgardnr committed:
Merge pull request #1275 from coralproject/flags-enh

Change Flags to Reports and make font a bit bigger

kgardnr committed:
Merge branch 'next' into account-history

kgardnr committed:
Merge pull request #1274 from coralproject/account-history

Suspension duration, fixes and wording for Account History

kgardnr committed:
Merge branch 'next' into email-send-fix

wyattjoh committed:
hides the user status histories and tokens from the auth endpoints, which were resulting in errors

wyattjoh committed:
Merge branch 'next' into hide-user-status-auth

wyattjoh committed:
allow moderators to view role

kgardnr committed:
Merge branch 'next' into perm-fix

wyattjoh committed:
added a plugin context

wyattjoh committed:
removed config from context

kgardnr committed:
Merge pull request #1267 from coralproject/perm-fix

Permission fix

okbel committed:
Merge branch 'next' into hide-user-status-auth

okbel committed:
Merge branch 'next' into email-send-fix

okbel committed:
Merge branch 'next' into plugin-context

wyattjoh committed:
fixing spelling :(

kgardnr committed:
Merge pull request #1277 from coralproject/hide-user-status-auth

Hide user status auth

kgardnr committed:
Merge branch 'next' into email-send-fix

wyattjoh committed:
fixed import typo

wyattjoh committed:
Merge branch 'next' into plugin-context

wyattjoh committed:
Merge pull request #1278 from coralproject/plugin-context

Plugin context

wyattjoh committed:
Merge branch 'next' into email-send-fix

wyattjoh committed:
Merge pull request #1276 from coralproject/email-send-fix

Missing Email Error

kgardnr committed:
Merge branch 'master' into next

wyattjoh committed:
fixed circle.yml

wyattjoh committed:
updated docs for 4.0

wyattjoh committed:
added deprecation warning

kgardnr committed:
Merge branch 'next' into docs

kgardnr committed:
Small wording tweak

wyattjoh committed:
Merge pull request #1279 from coralproject/docs

Updated docs for 4.0

wyattjoh committed:
replaced eslint:recommended with prettier
